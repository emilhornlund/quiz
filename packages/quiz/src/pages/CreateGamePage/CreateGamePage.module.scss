@import '../../styles/colors';
@import '../../styles/helpers';

.details {
  display: flex;
  flex-direction: row;
  column-gap: $layout-gutter;
}

.editor {
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 100%;
  flex: 1;
  position: relative;

  > textarea {
    padding-bottom: 2rem;
  }

  .error {
    position: absolute;
    right: 1rem;
    left: 1rem;
    bottom: 1rem;
    background-color: $red-2;
    padding: 0.5rem;
    border-radius: 0.5rem;
    color: $white;
    font-size: 14px;
    font-weight: bold;;
  }
}
