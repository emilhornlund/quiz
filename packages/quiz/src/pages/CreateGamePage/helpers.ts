import { CreateGameRequestDto, GameMode, QuestionType } from '@quiz/common'

type ClassicModeQuestions = Extract<
  CreateGameRequestDto,
  { mode: GameMode.Classic }
>['questions']

type ZeroToOneHundredModeQuestions = Extract<
  CreateGameRequestDto,
  { mode: GameMode.ZeroToOneHundred }
>['questions']

export type QuestionsForMode<T extends GameMode> = T extends GameMode.Classic
  ? ClassicModeQuestions
  : T extends GameMode.ZeroToOneHundred
    ? ZeroToOneHundredModeQuestions
    : never

/* eslint-disable-next-line */
const assertType = <T>(value: any, type: string, fieldName: string): T => {
  if (typeof value !== type) {
    throw new Error(
      `Invalid type for field '${fieldName}'. Expected ${type}, got ${typeof value}`,
    )
  }
  return value as T
}

export const parseQuestionsJson = <T extends GameMode>(
  /* eslint-disable-next-line */
  parsedJson: any[],
  mode: T,
): QuestionsForMode<T> => {
  if (mode === GameMode.Classic) {
    return parsedJson.map((question) => {
      if (question.type === QuestionType.Multi) {
        return {
          type: QuestionType.Multi,
          question: assertType<string>(question.question, 'string', 'question'),
          imageURL: assertType<string>(question.imageURL, 'string', 'imageURL'),
          /* eslint-disable-next-line */
          answers: question.answers.map((answer: any) => ({
            value: assertType<string>(answer.value, 'string', 'answers.value'),
            correct: assertType<boolean>(
              answer.correct,
              'boolean',
              'answers.correct',
            ),
          })),
          points: assertType<number>(question.points, 'number', 'points'),
          duration: assertType<number>(question.duration, 'number', 'duration'),
        } as ClassicModeQuestions[number]
      } else if (question.type === QuestionType.TrueFalse) {
        return {
          type: QuestionType.TrueFalse,
          question: assertType<string>(question.question, 'string', 'question'),
          imageURL: assertType<string>(question.imageURL, 'string', 'imageURL'),
          correct: assertType<boolean>(question.correct, 'boolean', 'correct'),
          points: assertType<number>(question.points, 'number', 'points'),
          duration: assertType<number>(question.duration, 'number', 'duration'),
        } as ClassicModeQuestions[number]
      } else if (question.type === QuestionType.Slider) {
        return {
          type: QuestionType.Slider,
          question: assertType<string>(question.question, 'string', 'question'),
          imageURL: assertType<string>(question.imageURL, 'string', 'imageURL'),
          min: assertType<number>(question.min, 'number', 'min'),
          max: assertType<number>(question.max, 'number', 'max'),
          correct: assertType<number>(question.correct, 'number', 'correct'),
          points: assertType<number>(question.points, 'number', 'points'),
          duration: assertType<number>(question.duration, 'number', 'duration'),
        } as ClassicModeQuestions[number]
      } else if (question.type === QuestionType.TypeAnswer) {
        return {
          type: QuestionType.TypeAnswer,
          question: assertType<string>(question.question, 'string', 'question'),
          imageURL: assertType<string>(question.imageURL, 'string', 'imageURL'),
          correct: assertType<string>(question.correct, 'string', 'correct'),
          points: assertType<number>(question.points, 'number', 'points'),
          duration: assertType<number>(question.duration, 'number', 'duration'),
        } as ClassicModeQuestions[number]
      } else {
        throw new Error('Unknown question type for Classic mode')
      }
    }) as QuestionsForMode<T>
  } else if (mode === GameMode.ZeroToOneHundred) {
    return parsedJson.map((question) => {
      if (question.type === QuestionType.Slider) {
        return {
          type: QuestionType.Slider,
          question: assertType<string>(question.question, 'string', 'question'),
          imageURL: assertType<string>(question.imageURL, 'string', 'imageURL'),
          correct: assertType<number>(question.correct, 'number', 'correct'),
          points: assertType<number>(question.points, 'number', 'points'),
          duration: assertType<number>(question.duration, 'number', 'duration'),
        } as ZeroToOneHundredModeQuestions[number]
      } else {
        throw new Error('Unknown question type for ZeroToOneHundred mode')
      }
    }) as QuestionsForMode<T>
  } else {
    throw new Error('Unsupported game mode')
  }
}
