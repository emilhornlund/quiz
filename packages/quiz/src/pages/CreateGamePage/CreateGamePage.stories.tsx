import type { Meta, StoryObj } from '@storybook/react'
import { withRouter } from 'storybook-addon-remix-react-router'

import CreateGamePage from './CreateGamePage'

const meta = {
  component: CreateGamePage,
  decorators: [withRouter],
  parameters: {
    layout: 'fullscreen',
  },
} satisfies Meta<typeof CreateGamePage>

export default meta
type Story = StoryObj<typeof meta>

export const Default = {
  args: {},
} satisfies Story
