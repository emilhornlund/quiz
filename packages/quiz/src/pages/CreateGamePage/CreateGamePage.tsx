import { CreateGameRequestDto, GameMode, QuestionType } from '@quiz/common'
import React, {
  FC,
  FormEvent,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'react'
import { useNavigate } from 'react-router-dom'

import {
  IconButtonArrowRight,
  Page,
  Select,
  Textarea,
  TextField,
} from '../../components'
import { useGameStorage } from '../../utils/use-game-storage'
import { useQuizService } from '../../utils/use-quiz-service.tsx'

import styles from './CreateGamePage.module.scss'
import { parseQuestionsJson, QuestionsForMode } from './helpers.ts'

const GameModeLabels: { [key in GameMode]: string } = {
  [GameMode.Classic]: 'Classic',
  [GameMode.ZeroToOneHundred]: '0 - 100',
}

const CreateGamePage: FC = () => {
  const navigate = useNavigate()

  const { createGame } = useQuizService()
  const [, persistGameData] = useGameStorage()

  const [name, setName] = useState<string>('')
  const [mode, setMode] = useState<GameMode>(GameMode.Classic)

  const [classicModeQuestions, setClassicModeQuestions] = useState<
    QuestionsForMode<GameMode.Classic>
  >([])
  const [zeroToOneHundredModeQuestions, setZeroToOneHundredModeQuestions] =
    useState<QuestionsForMode<GameMode.ZeroToOneHundred>>([])

  const [questionsJson, setQuestionsJson] = useState<string>('[]')
  const [jsonError, setJsonError] = useState<string>()

  const isValid = useMemo(() => {
    const isNameValid = !!name
    const isModeValid = !!mode
    const isQuestionsValid =
      !!(mode === GameMode.Classic && classicModeQuestions.length) ||
      !!(
        mode === GameMode.ZeroToOneHundred &&
        zeroToOneHundredModeQuestions.length
      )
    return isNameValid && isModeValid && isQuestionsValid
  }, [name, mode, classicModeQuestions, zeroToOneHundredModeQuestions])

  const handleChangeMode = useCallback((newMode: GameMode) => {
    setMode(newMode)
    if (newMode === GameMode.Classic) {
      handleParseQuestionJson(
        JSON.stringify(
          [
            {
              type: QuestionType.Multi,
              question: '',
              imageURL: '',
              answers: [
                { value: '', correct: false },
                { value: '', correct: false },
                { value: '', correct: false },
                { value: '', correct: false },
              ],
              points: 1000,
              duration: 30,
            },
            {
              type: QuestionType.TrueFalse,
              question: '',
              imageURL: '',
              correct: false,
              points: 1000,
              duration: 30,
            },
            {
              type: QuestionType.Slider,
              question: '',
              imageURL: '',
              min: 0,
              max: 100,
              correct: 0,
              points: 1000,
              duration: 30,
            },
            {
              type: QuestionType.TypeAnswer,
              question: '',
              imageURL: '',
              correct: '',
              points: 1000,
              duration: 30,
            },
          ],
          null,
          2,
        ),
        newMode,
      )
    }
    if (newMode === GameMode.ZeroToOneHundred) {
      handleParseQuestionJson(
        JSON.stringify(
          [
            {
              type: QuestionType.Slider,
              question: '',
              imageURL: '',
              correct: 0,
              points: 1000,
              duration: 30,
            },
          ],
          null,
          2,
        ),
        newMode,
      )
    }
  }, [])

  useEffect(() => {
    handleChangeMode(GameMode.Classic)
  }, [handleChangeMode])

  const handleChangeJSON = (value: string) => {
    handleParseQuestionJson(value, mode)
  }

  const handleParseQuestionJson = (value: string, mode: GameMode) => {
    setQuestionsJson(value)
    setJsonError(undefined)

    /* eslint-disable-next-line */
    let parsedJson: any
    let jsonValid = true
    try {
      parsedJson = JSON.parse(value)
    } catch (error) {
      jsonValid = false
      setJsonError((error as Error).message)
    }

    if (jsonValid) {
      if (mode === GameMode.Classic) {
        let parsedQuestions: QuestionsForMode<GameMode.Classic> = []
        try {
          parsedQuestions = parseQuestionsJson(parsedJson, mode)
        } catch (error) {
          jsonValid = false
          setJsonError((error as Error).message)
          setClassicModeQuestions([])
        }
        if (parsedQuestions && jsonValid) {
          setClassicModeQuestions(parsedQuestions)
        }
      }
      if (mode === GameMode.ZeroToOneHundred) {
        let parsedQuestions: QuestionsForMode<GameMode.ZeroToOneHundred> = []
        try {
          parsedQuestions = parseQuestionsJson(parsedJson, mode)
        } catch (error) {
          jsonValid = false
          setJsonError((error as Error).message)
          setZeroToOneHundredModeQuestions([])
        }
        if (parsedQuestions && jsonValid) {
          setZeroToOneHundredModeQuestions(parsedQuestions)
        }
      }
    }
  }

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault()

    if (isValid) {
      const request: CreateGameRequestDto = {
        name,
        mode: mode,
        questions: [],
      }
      if (mode === GameMode.Classic && classicModeQuestions.length) {
        request.questions = classicModeQuestions
      }
      if (
        mode === GameMode.ZeroToOneHundred &&
        zeroToOneHundredModeQuestions.length
      ) {
        request.questions = zeroToOneHundredModeQuestions
      }

      createGame(request)
        .then((response) => {
          persistGameData(response.id, response.token)
          navigate(`/game?gameID=${response.id}`)
        })
        .catch(console.error)
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      <Page
        header={
          <IconButtonArrowRight
            id="create-game-button"
            type="submit"
            kind="secondary"
            size="small"
            value="Create Game"
            disabled={!isValid}
          />
        }>
        <div className={styles.details}>
          <TextField
            id="name-textfield"
            type="text"
            value={name}
            placeholder="Name"
            onChange={setName}
          />
          <Select
            id="game-mode-select"
            values={Object.entries(GameModeLabels).map(([key, value]) => ({
              key,
              value: key,
              valueLabel: value,
            }))}
            value={mode}
            onChange={(value) => handleChangeMode(value as GameMode)}
          />
        </div>
        <div className={styles.editor}>
          <Textarea
            id="json"
            name="json"
            value={questionsJson}
            onChange={handleChangeJSON}
          />
          {jsonError && <div className={styles.error}>{jsonError}</div>}
        </div>
      </Page>
    </form>
  )
}

export default CreateGamePage
