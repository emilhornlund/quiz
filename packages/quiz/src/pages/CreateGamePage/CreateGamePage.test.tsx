import { render } from '@testing-library/react'
import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { describe, expect, it } from 'vitest'

import CreateGamePage from './CreateGamePage'

describe('CreateGamePage', () => {
  it('should render CreateGamePage', async () => {
    const { container } = render(
      <MemoryRouter>
        <CreateGamePage />
      </MemoryRouter>,
    )

    expect(container).toMatchSnapshot()
  })
})
