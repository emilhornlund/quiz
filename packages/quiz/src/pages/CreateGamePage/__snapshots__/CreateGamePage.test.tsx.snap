// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateGamePage > should render CreateGamePage 1`] = `
<div>
  <form>
    <div
      class="main"
    >
      <div
        class="header"
      >
        <div
          class="logo"
        >
          <span
            class="icon"
          />
          <span
            class="text"
          >
            Quiz
          </span>
        </div>
        <div
          class="side"
        >
          <div
            class="main small"
          >
            <button
              class="secondary"
              data-testid="test-create-game-button-button"
              disabled=""
              id="create-game-button"
              name="create-game-button"
              type="submit"
            >
              <span>
                Create Game
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-right "
                data-icon="arrow-right"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M438.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L338.8 224 32 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l306.7 0L233.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="content fullWidth centerAlign"
      >
        <div
          class="details"
        >
          <div
            class="main"
          >
            <input
              data-testid="test-name-textfield-textfield"
              id="name-textfield"
              name="name-textfield"
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="select"
          >
            <select
              data-testid="test-game-mode-select-select"
              id="game-mode-select"
              name="game-mode-select"
            >
              <option
                value="CLASSIC"
              >
                Classic
              </option>
              <option
                value="ZERO_TO_ONE_HUNDRED"
              >
                0 - 100
              </option>
            </select>
          </div>
        </div>
        <div
          class="editor"
        >
          <div
            class="container"
          >
            <textarea
              data-testid="test-json-textarea"
              id="json"
              name="json"
            >
              [
  {
    "type": "MULTI",
    "question": "",
    "imageURL": "",
    "answers": [
      {
        "value": "",
        "correct": false
      },
      {
        "value": "",
        "correct": false
      },
      {
        "value": "",
        "correct": false
      },
      {
        "value": "",
        "correct": false
      }
    ],
    "points": 1000,
    "duration": 30
  },
  {
    "type": "TRUE_FALSE",
    "question": "",
    "imageURL": "",
    "correct": false,
    "points": 1000,
    "duration": 30
  },
  {
    "type": "SLIDER",
    "question": "",
    "imageURL": "",
    "min": 0,
    "max": 100,
    "correct": 0,
    "points": 1000,
    "duration": 30
  },
  {
    "type": "TYPE_ANSWER",
    "question": "",
    "imageURL": "",
    "correct": "",
    "points": 1000,
    "duration": 30
  }
]
            </textarea>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
